### ISSUE TITLE
Lint calling `mem::swap` on a reference

### ISSUE THREAD

**Issue by pengowen123** [rust-lang/rust-clippy#1968]
It is possible to accidentally swap references instead of the data itself, just like it is possible to call `mem::drop` on a reference. It may be useful in some cases to swap references to something like slices, but the lint could be disabled in those cases or just be allow by default.

**Comment by oli-obk** [comment]
I'm not sure about this one. We should add it as allow by default and run it on some crates

**Comment by pengowen123** [comment]
I think limiting this lint to references that are temporaries, such as a function returning a reference, would fit most cases. For example, doing `mem::swap(&mut self.foo(), &mut self.bar())`, where `foo` and `bar` return references to something, should warn, because it can't do anything useful. However, this shouldn't warn:

```
let mut x = &5;
let mut y = &7;
mem::swap(&mut x, &mut y);
```
